### PR TITLE
fix(raiko): avoid duplicate image uploads

### PR DESCRIPTION
**Pull request by smtmfft**
The block proof ELF is approximately 8 MB, and is being uploaded every time. Using an initialization-once approach to avoid redundant uploads

### CODE REVIEW

**Review by johntaiko — APPROVED** [review]

File: provers/risc0/driver/src/bonsai.rs [hunk 1]
```diff
@@ -263,6 +265,30 @@ pub async fn cancel_proof(uuid: String) -> anyhow::Result<()> {
...
+            .get_or_try_init(|| {
+                let client = Client::from_env(risc0_zkvm::VERSION)?;
+                client.upload_img(&encoded_image_id, elf)?;
+
+                Ok(BonsaiClient {
+                    client: Arc::new(client),
+                })
+            })
+            .map_err(|e| BonsaiExecutionError::SdkFailure(e))?)
+    }
```
Comment (keroro520) [inline provers/risc0/driver/src/bonsai.rs:289]: Personally, I believe that using a static shared instance can be risky. For instance, if the shared client instance becomes broken due to a network issue, is it possible for the shared client instance to automatically recover once the network problem is resolved?

How about checking whether the image has been uploaded when creating a new client instance? [`get_image_upload_url`](https://github.com/risc0/risc0/blob/main/bonsai/sdk/src/lib.rs#L569-L586)

```rust
            let res = self
                .client
                .get(format!("{}/images/upload/{}", self.url, image_id))
                .send()
                .await?;

            if res.status() == 204 {
                return Ok(ImageExistsOpt::Exists);
            }
```

Comment (keroro520) [reply provers/risc0/driver/src/bonsai.rs:289]: Oh, I notice that [`upload_img`](https://github.com/risc0/risc0/blob/main/bonsai/sdk/src/lib.rs#L608-L611) will not re-upload image if the image has been already uploaded. If so, what issue we are facing?

**Review by keroro520 — CHANGES_REQUESTED** [review]

**Comment by smtmfft** [comment]
Oh, yes, risc0 checked the image id. I update sp1 logic and thought the risc0 maybe the same.
Please help to double check the sp1 part. I will revert risc0 changes.

**Comment by smtmfft** [comment]
seems sp1 can not easily bypass the upload, although only optimizing the setup part makes the process slightly better.

**Review by keroro520 — APPROVED** [review]
LGTM

**Comment by keroro520** [comment]
> seems sp1 can not easily bypass the upload, although only optimizing the setup part makes the process slightly better.

https://github.com/succinctlabs/sp1/pull/1875 implement the v2 sdk, which gracefully handle the image upload: [`register_program`](https://github.com/succinctlabs/sp1/blob/bfb0c6d8e045b5f40422b9c06cb0e9ee21b3c19c/crates/sdk/src/network/client.rs#L75)

We can switch to the new version of SDK once it becomes stable.

File: provers/risc0/driver/src/bonsai.rs (hunk 1 above)
Comment (smtmfft) [reply provers/risc0/driver/src/bonsai.rs]: I think your concerns is solid, here risc0 has no change, but sp1 side has a similar static instance, need to figure out a way to make sure the client is alive.

**Comment by smtmfft** [comment]
cool, so far we rely on their network service, let's see when they will update the service, probably in 4.0.0.

**Pull request merged**